docs/proposals/scheduling/521-replica-estimator/README.md

```diff
@@ -0,0 +1,179 @@
+
+Here's the architecture design diagram.
+
+![design](design.png)
+
+### API Changes
+
+First, the API of `ResourceBinding` must be changed. `NodeAffinity`, `NodeSelector` and `Tolerations` should be added as a representative for node claim along with the existed `ResourceRequest`.
+
+```go
+// ReplicaRequirements represents the requirements required by each replica.
+type ReplicaRequirements struct {
+	// A node selector represents the union of the results of one or more label queries
+	// over a set of nodes; that is, it represents the OR of the selectors represented
+	// by the node selector terms.
+	// +optional
+	NodeAffinity *corev1.NodeSelector `json:"nodeAffinity,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// ResourceRequest represents the resources required by each replica.
+	// +optional
+	ResourceRequest corev1.ResourceList `json:"resourceRequest,omitempty"`
+}
+```
+
+After the change, the `Object Reference` will be like this. The only difference is converting `corev1.ResourceList` to `*ReplicaRequirements` required by each replica.
+
+```go
+// ObjectReference contains enough information to locate the referenced object inside current cluster.
+type ObjectReference struct {
+	// APIVersion represents the API version of the referent.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind represents the Kind of the referent.
+	Kind string `json:"kind"`
+
+	// Namespace represents the namespace for the referent.
+	// For non-namespace scoped resources(e.g. 'ClusterRole')，do not need specify Namespace,
+	// and for namespace scoped resources, Namespace is required.
+	// If Namespace is not specified, means the resource is non-namespace scoped.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Name represents the name of the referent.
+	Name string `json:"name"`
+
+	// ResourceVersion represents the internal version of the referenced object, that can be used by clients to
+	// determine when object has changed.
+	// +optional
+	ResourceVersion string `json:"resourceVersion,omitempty"`
+
+	// ReplicaRequirements represents the requirements required by each replica.
+	// +optional
+	ReplicaRequirements *ReplicaRequirements `json:"replicaRequirements,omitempty"`
+
+	// Replicas represents the replica number of the referencing resource.
+	// +optional
+	Replicas int32 `json:"replicas,omitempty"`
```

Comment:
It would be better to add this fields in ResourceBingding.spec, but keep ObjectReference focusing on indicating the relationship between binding and resource.

Reply:
Good idea. Now we got `ReplicaResourceRequirements` in `ObjectReference`, which represents the resources required by each replica. Should we move `ReplicaResourceRequirements` into `ResourceBindingSpec` first in next PR and then modify it to `ReplicaRequirements`?

Reply:
> Should we move ReplicaResourceRequirements into ResourceBindingSpec first in next PR

What do you mean `in next PR`? I think we can do it before this proposal.

Both `ReplicaResourceRequirements` and `Replicas` should be moved out, right?
https://github.com/karmada-io/karmada/blob/4c2c18966d1f0e882d1a0655938fd96ab413ba19/pkg/apis/work/v1alpha1/binding_types.go#L59-L65

Reply:
>What do you mean in next PR? I think we can do it before this proposal.

My bad. I mean before this proposal.

>Both ReplicaResourceRequirements and Replicas should be moved out, right?

Yes.